libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/MetadataSyncManager.kt

```diff
@@ -0,0 +1,259 @@
+        syncCallback: MetadataSyncCallback
+    ) {
+        when (mode) {
+            Constants.Mode.CACHE_ONLY -> fetchFromCache(objectType, syncCallback, false)
+            Constants.Mode.CACHE_FIRST -> fetchFromCache(objectType, syncCallback, true)
+            Constants.Mode.SERVER_FIRST -> fetchFromServer(objectType, syncCallback)
+        }
+    }
+
+    init {
+        initializeSoup()
+    }
+
+    private fun fetchFromServer(objectType: String, syncCallback: MetadataSyncCallback) {
+        val target: SyncDownTarget = MetadataSyncDownTarget(objectType)
+        val options: SyncOptions = SyncOptions.optionsForSyncDown(MergeMode.OVERWRITE)
+        try {
+            syncManager.syncDown(target, options, SOUP_NAME, object:SyncManager.SyncUpdateCallback {
+                override fun onUpdate(sync: SyncState) {
+                    if (SyncState.Status.DONE == sync.status) {
+                        fetchFromCache(objectType, syncCallback, false)
+                    }
+                }
+            })
+        } catch (e: Exception) {
+            MobileSyncLogger.e(TAG, "Exception occurred while reading metadata from the server", e)
+        }
+    }
+
+    private fun fetchFromCache(
+        objectType: String, syncCallback: MetadataSyncCallback,
+        fallbackOnServer: Boolean
+    ) {
+        val querySpec = QuerySpec.buildSmartQuerySpec(
+            String.format(
+                QUERY,
+                objectType
+            ), 1
+        )
+        try {
+            val results = smartStore.query(querySpec, 0)
+            if ((results == null) || (results.length() == 0)) {
+                if (fallbackOnServer) {
+                    fetchFromServer(objectType, syncCallback)
+                } else {
+                    onSyncComplete(syncCallback, null)
+                }
+            } else {
+                onSyncComplete(
+                    syncCallback,
+                    Metadata.fromJSON(results.optJSONArray(0).optJSONObject(0))
+                )
+            }
+        } catch (e: Exception) {
+            MobileSyncLogger.e(TAG, "Exception occurred while reading metadata from the cache", e)
+        }
+    }
+
+    private fun onSyncComplete(syncCallback: MetadataSyncCallback?, metadata: Metadata?) {
+        syncCallback?.onSyncComplete(metadata)
+    }
+
+    private fun initializeSoup() {
+        if (!smartStore.hasSoup(SOUP_NAME)) {
+            smartStore.registerSoup(SOUP_NAME, INDEX_SPECS)
+        }
+    }
+
+    /**
+     * Callback interface for metadata sync.
+     *
+     * @author bhariharan
+     */
+    interface MetadataSyncCallback {
+        /**
+         * Callback triggered when metadata sync completes.
+         *
+         * @param metadata Metadata.
+         */
+        fun onSyncComplete(metadata: Metadata?)
+    }
+
+    companion object {
+        const val SOUP_NAME = "sfdcMetadata"
+        const val QUERY = "SELECT {" + SOUP_NAME + ":_soup} FROM {" + SOUP_NAME +
```

Comment:
Optional: Here's another string template conversion that'd be a quick win.  It might be worth looking across the whole module for them.